### DISCUSSION TITLE
Alternative extensions syntax related to LDM October 14th, 2024

### DISCUSSION THREAD

**Discussion by OJacot-Descombes** [dotnet/csharplang#8524] [LDM Notes]
What about overloaded extensions, i.e., extensions having the same name, but for different types?

Instead of the syntax proposed in the [LDM for October 14th, 2024](https://github.com/dotnet/csharplang/blob/main/meetings/2024/LDM-2024-10-14.md)
``` c#
extension E
{
    for IEnumerable<T> where T
    {
        public void M1() { ... }
    }

    for IList<T> where T
    {
        public void M2() { ... }
    }
}
```
we would have

``` c#
extension E for IEnumerable<T> where T
{
    public void M1() { ... }
}

extension E for IList<T> where T
{
    public void M2() { ... }
}
```
And we could introduce the type parameter like this:
``` c#
extension E<T> for IEnumerable<T> where T
{
    public void M1() { ... }
}

extension E<T> for IList<T> where T
{
    public void M2() { ... }
}
```

**Comment by HaloFour** [comment]
Similar spaghetti at the wall, partial extensions that each can have a different target.  If combined with flattening the generics (e.g. the extension generic type parameters are prepended to the extension members), that would also make them binary compatible with existing extension methods.

```cs
public partial extension Enumerable for IEnumerable {
    public IEnumerable<T> OfType<T>() { ... }
}

public partial extension Enumerable<T> for IEnumerable<T> {
    public IEnumerable<T> Where(Func<T, bool> predicate) { ... }
}

public partial extension Enumerable<T> for IOrderedEnumerable<T> {
    public IOrderedEnumerable<T> ThenBy<TKey>(Func<T, TKey> selector) { ... }
}

public partial static class Enumerable {
    public static IEnumerable<T> Empty<T>() { ... }
}
```

**Comment by 333fred** [comment]
This version of the syntax did come up during the meeting. It still has the problem of "where is the type parameter introduced", so we're not satisfied with it.

**Reply by laicasaane** [reply]
What does it mean "where is the type parameter introduced"? Would you kindly explain it to me? I can't see the problem in that proposed alternative. Thanks.

**Reply by CyrusNajmabadi** [reply]
> `extension E for IEnumerable<T> where T`

Both `IEnumerable<T>` and `where T` refer to this symbol T. But where was T declared?

**Reply by laicasaane** [reply]
There is no way for the language to know that T is from `IEnumerable<T>`?

**Reply by yaakov-h** [reply]
Consider the following:

```csharp
using System.Collections.Generic;

public class T { }

public static class Extensions
{
    public static void Foo<T>(this IEnumerable<T> o) => throw null!;
    
    public static void Bar(this IEnumerable<T> o) => throw null!;
}
```

The first method is generic, the second is not. How can you tell if `IEnumerable<T>` is a generic or not if we don't declare `T` within the definition of the extension?

**Reply by laicasaane** [reply]
That makes sense. Thanks!

**Reply by OJacot-Descombes** [reply]
I updated my suggestion. We could write `extension E<T> for IEnumerable<T> where T` by declaring the type parameter after the extension name, just as we do for classes.

**Reply by jnm2** [reply]
@CyrusNajmabadi I had originally suggested as a strawman that `where T` could declare T

**Reply by CyrusNajmabadi** [reply]
> @CyrusNajmabadi I had originally suggested as a strawman that `where T` could declare T

Right. That might be a solution to the "where is the type parameter declared" issue. But I def think it's not an obvious thing.

**Reply by jnm2** [reply]
Just dreaming, but associated type declaration syntax could move the location of the T declaration in a way that comes together nicely with this.

**Reply by CyrusNajmabadi** [reply]
> Just dreaming, but associated type declaration syntax could move the location of the T declaration in a way that comes together nicely with this.

What do you think that would look like here? :-)

**Reply by jnm2** [reply]
Something I'd have to sit on. But `where` could morph into the declaration line for the associated type, maybe on the other side of the curly brace. e.g. `class T : IEnumerable<T>;` is valid today and introduces a nested type, and `where T : IEnumerable<T>;` could then introduce an associated type. (Same with `class T;` and `where T;`)

**Reply by Richiban** [reply]
We could take inspiration from (I think) Kotlin, and have something like `extension<T> E for IEnumerable<T> { ... }`
